scripts/run-registry.ts

```diff
@@ -130,55 +145,88 @@ const currentVersion = async () => {
-   * yarn specific version "ranges". such as "workspace:*"
-   *
-   * We can't publish to npm if the package.json contains these ranges. So with `yarn pack` we
-   * create a tarball that we can publish.
-   *
-   * However this bug exists in NPM: https://github.com/npm/cli/issues/4533! Which causes the NPM
-   * CLI to disregard the tarball CLI argument and instead re-create a tarball. But NPM doesn't
-   * replace the yarn version ranges.
-   *
-   * So we create the tarball ourselves and move it to another location on the FS. Then we
-   * change-directory to that directory and publish the tarball from there.
-   */
-  await mkdir(PACKS_DIRECTORY, { recursive: true }).catch(() => {});
-
-  return Promise.all(
-    packages.map(({ name, location }) =>
-      limit(
-        () =>
-          new Promise((resolve, reject) => {
-            const loggedLocation = location.replace(resolvePath(join(__dirname, '..')), '.');
-            const resolvedLocation = resolvePath('../code', location);
-
-            logger.log(`🛫 publishing ${name} (${loggedLocation})`);
-
-            const tarballFilename = `${name.replace('@', '').replace('/', '-')}.tgz`;
-            const command = `cd "${resolvedLocation}" && yarn pack --out="${PACKS_DIRECTORY}/${tarballFilename}" && cd "${PACKS_DIRECTORY}" && npm publish "./${tarballFilename}" --registry ${url} --force --tag="xyz" --ignore-scripts`;
-            exec(command, (e) => {
-              if (e) {
-                reject(e);
-              } else {
-                i += 1;
-                logger.log(`${i}/${packages.length} 🛬 successful publish of ${name}!`);
-                resolve(undefined);
-              }
-            });
-          })
-      )
-    )
-  );
+interface Package {
+  name: string;
+  location: string;
+  path: string;
+  version: string;
+  publishVersion: string;
+}
+
+const publishPackage = async (ws: Package, workspaces: Package[]) => {
+  const info = workspaces.find((it) => it.name === ws.name);
+  const tmpDir = await mkdtemp(join(tmpdir(), ws.name.replace('/', '-') + '-'));
+
+  await cp(ws.path, tmpDir, { recursive: true, force: true });
+
+  const pkg = JSON.parse(await readFile(join(tmpDir, 'package.json'), 'utf8'));
+  pkg.version = info.publishVersion;
+  resolveWorkspaceDeps(pkg, workspaces);
+
+  await writeFile(join(tmpDir, 'package.json'), JSON.stringify(pkg, null, 2));
+  await writeFile(join(tmpDir, '.npmrc'), `//localhost:${VERDACCIO_PORT}/:_authToken=fake`);
+
+  try {
+    await x(
+      'npm',
+      [
+        'publish',
+        '--registry',
+        `http://localhost:${VERDACCIO_PORT}`,
+        '--tag',
+        PUBLISH_TAG,
+        '--ignore-scripts',
+      ],
+      { nodeOptions: { cwd: tmpDir, stdio: 'inherit' }, throwOnError: true }
+    );
+  } finally {
+    await rm(tmpDir, { recursive: true, force: true }).catch(() => {});
+  }
 };
 
+// Resolve workspace dependencies to the published version
```

Comment:
the comment you deleted was very helpful. I understand that the flow is now different, but it essentially solves the same thing just in a different way, so I would expect an equally verbose comment to explain why.